### PR TITLE
API 2.0

### PR DESCRIPTION
**Pull request by avive**
Draft PR to review and discuss GlobalState service changes to support Spacemesh Apps in release 0.3.

### CODE REVIEW

File: proto/spacemesh/v1/app_types.proto [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    repeated bytes data = 2; // One item - primitive var data. More than one item - an array of primitive data.
+}
+
+// The address of an app template in the templates global state store
+message AppTemplateAddress {
+    bytes address = 1;
+}
+
+// Information about an app template in the global state store
+message AppTemplate {
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:42]: I think we can make it `Template`

File: proto/spacemesh/v1/app_types.proto [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// The meta-data and value of an app's variable
+message Variable {
+    VariableMetadata meta_data = 1;
+    repeated bytes data = 2; // One item - primitive var data. More than one item - an array of primitive data.
+}
+
+// The address of an app template in the templates global state store
+message AppTemplateAddress {
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:37]: In the SVM code there is no `AppTemplateXXX` - so I'd rename it to `TemplateAddress`

File: proto/spacemesh/v1/app_types.proto [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    VARIABLE_TYPE_AMOUNT = 10;
+    VARIABLE_TYPE_ADDRESS = 11;
+}
+
+// Variable metadata
+message VariableMetadata {
+    uint32 id = 1; // var index in app's state
+    string name = 2; // var name
+    VariableType type = 3; // type
+    bool is_array = 4; // when true, variable is an array of 0 or more primitive value types, otherwise, a single primitive type value.
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:27]: I'm not well versed in Profobuf - but if it's possible to specify that a variable is either a primitive or an Array of primitives I would do that.

File: proto/spacemesh/v1/global_state.proto [hunk 4]
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
...
+    rpc GetAppTemplate(GetAppTemplateRequest) returns (GetAppTemplateResponse);
+
+    // Get all app templates
+    rpc GetAllAppTemplates(GetAllAppTemplatesRequest) returns (GetAllAppTemplatesResponse);
+
+    // Get app metadata and optionally current state
+    rpc GetApp(GetAppRequest) returns (GetAppResponse);
+
+    // Get an app's variable metadata and optionally value
+    rpc GetAppVariable(GetAppVariableRequest) returns (GetAppVariableResponse);
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/global_state.proto:25]: As I've said - I wouldn't add it right now.
Probably it won't be used when we have only fixed-sized variables.

File: proto/spacemesh/v1/global_state.proto [hunk 5]
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
...
 // Readonly global state data - current and historical.
 // Global state data is data which is not explicitly stored in the mesh.
 // Global state is modified only by the state transition function.
 service GlobalStateService {
+
     // Latest computed global state - layer and its root hash
     rpc GlobalStateHash(GlobalStateHashRequest) returns (GlobalStateHashResponse);
 
+    // Get an app template meta-data and code
+    rpc GetAppTemplate(GetAppTemplateRequest) returns (GetAppTemplateResponse);
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/global_state.proto:16]: `GetTemplate` / `GetTemplateResponse` ?

File: proto/spacemesh/v1/global_state.proto [hunk 6]
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
...
+
+    // Get all app templates
+    rpc GetAllAppTemplates(GetAllAppTemplatesRequest) returns (GetAllAppTemplatesResponse);
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/global_state.proto:19]: Since we want to support pagination here, I prefer to call it `GetTemplates`, or maybe `TemplateIterator`

for the response type - how about `GetTemplatesResponse` ?

File: proto/spacemesh/v1/app_types.proto [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+// A request to return a template by address from the template global state store
+message GetAppTemplateRequest {
+    AppTemplateAddress address = 1; // template address
+    bool include_code = 2; // when true, template result will include full wasm code
+}
+
+// A response with AppTemplate data
+message GetAppTemplateResponse {
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:79]: `GetTemplateResponse` ?

File: proto/spacemesh/v1/app_types.proto [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    string name = 2; // short descriptive name
+    AppTemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
+    TransactionId spawn_tx_id = 5; // Transaction which spawned the app. empty for genesis apps.
+    AccountId spawner = 6; // App spawner account
+    repeated Variable variables = 8; // ordered array of app vars and their values
+}
+
+// A request to return a template by address from the template global state store
+message GetAppTemplateRequest {
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:73]: `GetTemplateResquest` ?

File: proto/spacemesh/v1/app_types.proto [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+// A response with variable meta-data and value
+message GetAppVariableResponse {
+    Variable variable_value = 1;
+}
+
+// A request to get a deployed app metadata and optionally all variable values as of a specific layer or latest state.
+message GetAppRequest {
+    AccountId address = 1; // app address
+    LayerNumber layer_number = 2; // layer for state. Ignored when include_var_values is set to false. pass MAX_UI32 to get latest state.
+    bool include_var_values = 3; // when true, var values will be included in response, false otherwise.
```
Comment (YaronWittenstein) [inline proto/spacemesh/v1/app_types.proto:112]: I didn't add that option under the issue - but I do think it makes sense to have it

File: proto/spacemesh/v1/app_types.proto (hunk 3 above)
Comment (YaronWittenstein) [reply proto/spacemesh/v1/app_types.proto:27]: In the future, there might be more ways to represent non-primitive variables...

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]

File: proto/spacemesh/v1/app_types.proto (hunk 1 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto]: Done - and same everywhere

File: proto/spacemesh/v1/app_types.proto (hunk 2 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto]: done

File: proto/spacemesh/v1/global_state.proto (hunk 4 above)
Comment (avive) [reply proto/spacemesh/v1/global_state.proto]: let's talk about - i think it will be very useful for api clients even with fixed-size vars

File: proto/spacemesh/v1/app_types.proto (hunk 3 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:28]: done

File: proto/spacemesh/v1/app_types.proto [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    bool is_array = 4; // when true, variable is an array of 0 or more primitive value types, otherwise, a single primitive type value.
+}
+
+message ValuesArray {
+    repeated bytes values = 1;
+}
+
+// The meta-data and value of an app's variable
+message Variable {
+    VariableMetadata meta_data = 1;
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:37]:
```suggestion
    VariableMetadata metadata = 1;
```

File: proto/spacemesh/v1/app_types.proto [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    string name = 2; // var name
+    VariableType type = 3; // type
+    bool is_array = 4; // when true, variable is an array of 0 or more primitive value types, otherwise, a single primitive type value.
+}
+
+message ValuesArray {
+    repeated bytes values = 1;
+}
+
+// The meta-data and value of an app's variable
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:35]:
```suggestion
// The metadata and value of a state value
```

File: proto/spacemesh/v1/app_types.proto [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+message Variable {
+    VariableMetadata meta_data = 1;
+    oneof data {
+        bytes value = 2; // since bytes size is arbitrary - this will support future var types
+        ValuesArray values = 3;
+    }
+}
+
+// The address of an app template in the templates global state store
+message TemplateAddress {
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:45]: Why do we need a distinct type for this? Why not use the existing `Address` type?

File: proto/spacemesh/v1/app_types.proto [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    bytes address = 1;
+}
+
+// Information about an app template in the global state store
+message Template {
+    TemplateAddress address = 1; // template address
+    string name = 2; // template name
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:52]: See comment in issue, I don't understand why this is required

File: proto/spacemesh/v1/app_types.proto [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+    // note - the author can be obtained from the deploy transaction but it is nice to have it here
+    AccountId author = 3; // template uploader
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:55]:
```suggestion
    AccountId deployer = 3; // template uploader
```

File: proto/spacemesh/v1/global_state_types.proto [hunk 15]
```diff
@@ -50,22 +50,42 @@ message AccountDataQueryRequest {
     uint32 offset = 3; // return results from offset
 }
 
+message ExecutionError {
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:53]: We have standard error/status types already, with the same data elements, maybe we don't need to introduce another one here. See "google.golang.org/grpc/codes", "google.golang.org/grpc/status"

File: proto/spacemesh/v1/app_types.proto [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    bytes code = 6; // template wasm code
+    repeated VariableMetadata vars = 7; // template variables order and metadata
+}
+
+// A Spacemesh app
+// Note that both layer and spawner can be obtained from the transaction and transaction id for apps spawned by a transaction
+// but is convenient to return them directly as part of apps' metadata.
+message App {
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:69]: Do we need this?

File: proto/spacemesh/v1/app_types.proto [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    TemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:71]: Why is this needed?

File: proto/spacemesh/v1/app_types.proto [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    bool include_code = 3; // when try templates results will include full wasm code
+}
+
+// A response includes total templates available in store and a list of zero ore more templates
+message GetTemplatesResponse {
+    uint32 total_results = 1; // total number of templates
+    repeated Template templates = 2;
+}
+
+// A request to get an app's variable by var id
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:101]: Instead of "variable" can we give these a more descriptive name, e.g., state entry?

File: proto/spacemesh/v1/app_types.proto [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+message GetVariableRequest {
+    AccountId address = 1; // app's address
+    LayerNumber layer_number = 2; // requested value as of a layer. Pass MAX_UI32 to get latest value
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:104]: As noted in the issue, will only work for archival nodes! What do non-archival nodes do if this is specified? Error?

File: proto/spacemesh/v1/app_types.proto [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+    uint32 id = 3; // app's var id - var's unique identifier
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:105]: Where does the caller look this up?

**Review by lrettig — CHANGES_REQUESTED** [review]

File: proto/spacemesh/v1/app_types.proto (hunk 20 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:105]: Caller has the app's schema which includes the variable ids.

File: proto/spacemesh/v1/app_types.proto (hunk 19 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:104]: See my comment there - full nodes will hold state for n layers back and archive nodes since genesis but from an api level it is good to have this like this and nodes will return not_found status if they can't satisfy this specific request for a layer which is not recent.

File: proto/spacemesh/v1/app_types.proto (hunk 17 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:71]: Which field? These are all meta-data about an app which is useful. It is useful to know in which layer an app was deployed to the mesh although this info can be obtained from the spawn transaction - it is just convince to include some metadata that can be obtained from transactions about the app in response to this call because the intent of the caller is to get this data.

File: proto/spacemesh/v1/app_types.proto (hunk 16 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:69]: I think it is good to have a short app name that describes it. I think that it is something that is set by spwaner and spawn time per app.

File: proto/spacemesh/v1/app_types.proto (hunk 14 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:55]: I'm using the svm terminology as specified by yaron, until we discuss making any changes to this terminology with yaron.

File: proto/spacemesh/v1/app_types.proto (hunk 12 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:45]: I think because in current SVM design, templates are stored in their own namespace which is not just an abstracted account address. We reviewed this as far as I remember in recent svm calls. Yaron can clarify.

File: proto/spacemesh/v1/app_types.proto (hunk 11 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto]: done

File: proto/spacemesh/v1/app_types.proto (hunk 10 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto]: done

File: proto/spacemesh/v1/app_types.proto (hunk 12 above)
Comment (avive) [reply proto/spacemesh/v1/app_types.proto:45]: Item for design review zoom

**Comment by avive** [comment]
@lrettig @YaronWittenstein - I've updated the api to reflect the design review session feedback. I still believe we should have a strongly typed GetVariable() method which is very useful for quick access to app state by dapps. Perhaps we can proceed before another review session by you guys just reviewing the most recent changes? The relent types are in app_types.proto, var_types.proto and global_state.proto

File: proto/spacemesh/v1/api_config.yaml [hunk 21]
```diff
@@ -103,7 +115,7 @@ http:
       post: "/v1/smesher/createpostdata"
       body: "*"
     - selector: spacemesh.v1.SmesherService.StopPostDataCreationSession
-      post: "/v1/smesher/stoppostdatacreationsession"
+      post: "/v1/smesher/stoppost:datacreationsession"
```
Comment (lrettig) [inline proto/spacemesh/v1/api_config.yaml:118]: is the `:` a typo?

File: proto/spacemesh/v1/app_types.proto (hunk 12 above)
Comment (lrettig) [reply proto/spacemesh/v1/app_types.proto:10]: then, rather than introducing a new type, maybe add a `namespace` to the existing `Address` type

File: proto/spacemesh/v1/app_types.proto [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+message TemplateAddress {
+    bytes address = 1;
+}
+
+// Template deploy data
+message DeployData {
+    TemplateAddress address = 1; // template address
+    // note - the author can be obtained from the deploy transaction but it is nice to have it here
+    AccountId author = 2; // template uploader
+    LayerNumber deploy_layer = 3; // layer that the template was deployed - 0 for genesis templates
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:19]: is this actually stored in the mesh data for a template? is it necessary for SVM? (both layer, and txid)

File: proto/spacemesh/v1/app_types.proto [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    TransactionId deploy_tx_id = 4; // the id of the deploy template transaction. Empty for genesis templates.
+}
+
+// Template code
+message Code {
+    string target_svm_version = 4; // at deploy time
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:25]: Can we collapse target svm version and kind down into one field? Can't it be included in the code itself as a "magic word" (first few bytes)?

File: proto/spacemesh/v1/app_types.proto [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+    string kind = 5; // e.g. WASM
+    repeated bytes code = 6; // wasm code
+}
+
+// Information about a template in the global state store
+message Template {
+    DeployData deploy_data = 1; // always provided.
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:32]: is this included in mesh data? is it required?

File: proto/spacemesh/v1/app_types.proto [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+// A response with AppTemplate data
+message GetTemplateResponse {
+    Template template = 1;
+}
+
+// A request to return all templates from the templates global state store with pagination
+message GetTemplatesRequest {
+    uint32 max_results = 1; // max number of results to return
+    uint32 offset = 2; // query offset
+    bool include_code = 3; // include code in response
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:58]: why not just always include both code and metadata? is it any additional work on the backend? it's not that much additional data

File: proto/spacemesh/v1/app_types.proto [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+    bool include_meta_data = 4; // include metadata in response if provided by author
+}
+
+// A response includes total templates available in store and a list of zero ore more templates
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:62]:
```suggestion
// A response includes total templates available in store and a list of zero or more templates
```

File: proto/spacemesh/v1/app_types.proto [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+///////////// App API
+
+// A Spacemesh app
+// Note that both layer and spawner can be obtained from the transaction and transaction id for apps spawned by a transaction
+// but is convenient to return them directly as part of apps' metadata.
+message App {
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name (optional)
+    TemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:78]: as with templates, above: are spawn layer, spawn txid, and spawner really needed for SVM/stored in mesh?

File: proto/spacemesh/v1/app_types.proto (hunk 19 above)
Comment (lrettig) [reply proto/spacemesh/v1/app_types.proto:87]: I think that works, but for the record, we cannot force full nodes to hold intermediate state for any minimum number of layers since this is extraprotocol

File: proto/spacemesh/v1/global_state_types.proto [hunk 28]
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
...
+}
+
+// From dev perspective we want to enable something like the following from the higher-level lang or rust code:
+// info([enty_name], [entry_text])
+// e.g. info("sig_verification", format!("%s : %d", str1, ui64))
+message LogEntry {
+    uint32 log_level = 1; // info, debug, trace, error, etc...
+    string context = 2; // function name e.g. withdraw()
+    string name = 3; // entry name
+    string entry = 4; // entry printed data
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:65]: in EVM/solidity, pretty sure this is a repeated set of field names and values rather than a single blob (per context/name)

File: proto/spacemesh/v1/global_state_types.proto [hunk 29]
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
...
     enum TransactionResult { // the results of STF transaction processing
         TRANSACTION_RESULT_UNSPECIFIED = 0;
         TRANSACTION_RESULT_EXECUTED = 1; // executed w/o error by the STF
         TRANSACTION_RESULT_BAD_COUNTER = 2; // unexpected transaction counter
-        TRANSACTION_RESULT_RUNTIME_EXCEPTION = 3; // app code exception
+        TRANSACTION_RESULT_RUNTIME_EXCEPTION = 3; // app code execution exception
         TRANSACTION_RESULT_INSUFFICIENT_GAS = 4; // out of gas
         TRANSACTION_RESULT_INSUFFICIENT_FUNDS = 5; // failed due to sender's insufficient funds
+        TRANSACTION_RESULT_APP_NOT_FOUND = 6; // an attempt to execute an app where there's no app at the app address
+        TRANSACTION_RESULT_FUNC_NOT_FOUND = 8; // called function not found in app
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:78]: is this distinguished, at the SVM level, from any other simple execution error? how/why?

File: proto/spacemesh/v1/global_state_types.proto [hunk 30]
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
...
+        TRANSACTION_RESULT_INSTANTIATION_FAILED = 9; // error creating an instance from a template
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:79]: call this spawn rather than instantiation for consistency

File: proto/spacemesh/v1/global_state_types.proto [hunk 31]
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
...
+        TRANSACTION_RESULT_DEPLOY_FAILED = 10; // error deploying a template to the mesh
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:80]: not sure we need to break out these two failure types for spawn/deploy since it should be clear which it is from context (tx type), no?

File: proto/spacemesh/v1/global_state_types.proto [hunk 32]
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
...
+
     uint64 gas_used = 3; // gas units used by the transaction
     Amount fee = 4; // transaction fee charged for the transaction (in smidge, gas_price * gas_used)
     LayerNumber layer = 5; // the layer in which the STF processed this transaction
-    uint32 index = 6; // the index of the tx in the ordered list of txs to be executed by stf in the layer.
-    bytes svm_data = 7; // svm binary data. Decode with svm-codec
+    uint32 index = 6; // the index of the tx in the ordered list of txs to be executed by stf in the layer
+
+    bytes svm_data = 7; // SVM data returned from an SVM function. Needs to be decoded with svm-codec
+    bytes app_state = 8; // when app function executed or app deployed, the app root state hash after execution has completed
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:90]: please call this `state_root` or something to be clearer

File: proto/spacemesh/v1/types.proto [hunk 33]
```diff
@@ -74,6 +74,7 @@ message Activation {
     AccountId coinbase = 4; // coinbase account id
     ActivationId prev_atx = 5; // previous ATX pointed to
     uint64 commitment_size = 6; // commitment size in bytes
+    EpochNumber epoch = 7; // Activation on mesh epoch number (when layer is not used)
```
Comment (lrettig) [inline proto/spacemesh/v1/types.proto:77]: I don't understand the comment here, when is "layer not used"?

File: proto/spacemesh/v1/types.proto [hunk 34]
```diff
@@ -110,9 +111,11 @@ message Block {
 message Layer {
     LayerNumber number = 1; // layer number - not hash - layer content may change
     enum LayerStatus {
-        LAYER_STATUS_UNSPECIFIED = 0; // not yet approved or confirmed
+        LAYER_STATUS_UNSPECIFIED = 0;
         LAYER_STATUS_APPROVED = 1; // approved by hare
         LAYER_STATUS_CONFIRMED = 2; // confirmed by tortoise
+        LAYER_STATUS_PENDING = 3; // not yet approved, confirmed or verified
```
Comment (lrettig) [inline proto/spacemesh/v1/types.proto:117]: shouldn't pending come before approved and confirmed?

File: proto/spacemesh/v1/var_types.proto [hunk 35]
```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
```
Comment (lrettig) [inline proto/spacemesh/v1/var_types.proto:1]: per our conversation last week, I don't think any of this belongs in the L1 API. I think it belongs at the application layer, and that this data should be returned from this API as a blob of "SVM compatible data"

**Review by lrettig — CHANGES_REQUESTED** [review]

**Comment by avive** [comment]
@YaronWittenstein @lrettig - please check the code - I think I've incorporated everything we talked about and removed non-app related changes from this pr.

File: proto/spacemesh/v1/app_types.proto [hunk 36]
```diff
@@ -2,47 +2,32 @@ syntax = "proto3";
...
+    repeated bytes code = 3; // wasm code
 }
 
 // Information about a template in the global state store
 message Template {
-    DeployData deploy_data = 1; // always provided.
-    Code code = 2; // template wasm code. Provided if caller specified to get code.
-    bytes metadata = 3; // storage layout, api, header. Provided if caller specified to get metadata and template include metadata.
+    Code code = 1; // template wasm code. Provided if caller specified to get code.
+    bytes metadata = 2; // storage layout, api, header. Provided if caller specified to get metadata and template include metadata.
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:18]:
```suggestion
    bytes metadata = 2; // storage layout, api, header. Provided if caller specified to get metadata and template including metadata. Caller is responsible for decoding.
```

File: proto/spacemesh/v1/app_types.proto [hunk 37]
```diff
@@ -2,47 +2,32 @@ syntax = "proto3";
...
 }
 
 enum TemplateDataFlags {
     TEMPLATE_DATA_FLAGS_UNSPECIFIED = 0;
     TEMPLATE_DATA_FLAGS_CODE = 1;
     TEMPLATE_DATA_FLAGS_METADATA = 2;
+    // may need deploy data section - tbd - based on deploy tx format. Understand eth one.
```
Comment (lrettig) [inline proto/spacemesh/v1/app_types.proto:25]:
```suggestion
    // TODO: may need deploy data section - tbd - based on deploy tx format. Understand eth one.
```

File: proto/spacemesh/v1/app_types.proto (hunk 27 above)
Comment (lrettig) [reply proto/spacemesh/v1/app_types.proto:62]: as discussed today, these also belong in metadata, as they're not required by SVM. they can optionally be automatically populated at deploy time if the deployer requests this, and pays for it.

File: proto/spacemesh/v1/app_types.proto (hunk 18 above)
Comment (lrettig) [reply proto/spacemesh/v1/app_types.proto:68]: @YaronWittenstein said these are called "storage variables" in SVM

File: proto/spacemesh/v1/global_state_types.proto [hunk 38]
```diff
@@ -70,14 +70,10 @@ message TransactionReceipt {
...
-        TRANSACTION_RESULT_BAD_COUNTER = 2; // unexpected transaction counter
-        TRANSACTION_RESULT_RUNTIME_EXCEPTION = 3; // app code execution exception
-        TRANSACTION_RESULT_INSUFFICIENT_GAS = 4; // out of gas
-        TRANSACTION_RESULT_INSUFFICIENT_FUNDS = 5; // failed due to sender's insufficient funds
-        TRANSACTION_RESULT_APP_NOT_FOUND = 6; // an attempt to execute an app where there's no app at the app address
-        TRANSACTION_RESULT_FUNC_NOT_FOUND = 8; // called function not found in app
-        TRANSACTION_RESULT_INSTANTIATION_FAILED = 9; // error creating an instance from a template
-        TRANSACTION_RESULT_DEPLOY_FAILED = 10; // error deploying a template to the mesh
+        TRANSACTION_RESULT_RUNTIME_EXCEPTION = 2; // app code execution exception=
+        TRANSACTION_RESULT_VALIDATION_FAILED = 3; // TBD
```
Comment (lrettig) [inline proto/spacemesh/v1/global_state_types.proto:74]:
```suggestion
        TRANSACTION_RESULT_VALIDATION_FAILED = 3; // TODO: do we need this? can an invalid tx ever make it into a superblock?
```

**Review by lrettig — CHANGES_REQUESTED** [review]

**Comment by avive** [comment]
Renamed pr to api 2.0 as it is not going to be backward compatible with 1.x api due to some breaking changes.

**Comment by lrettig** [comment]
> Renamed pr to api 2.0 as it is not going to be backward compatible with 1.x api due to some breaking changes.

I think we need to update this so the output proto files go into `proto/spacemesh/v2` (rather than `v1`), that should cause the CI tests to pass. We also need to add a separate set of checks for `v2`, and we should also update the README accordingly.
